### PR TITLE
enclosures are now strongly typed and are no longer vals

### PR DESCRIPTION
**Pull request by xeno-by**
review @paulp @retronym

### CODE REVIEW

**Comment by xeno-by** [comment]
Spurious failure of a performance test

**Comment by xeno-by** [comment]
PLS REBUILD ALL

File: src/compiler/scala/reflect/macros/runtime/Enclosures.scala [hunk 1]
```diff
@@ -1,22 +1,31 @@
...
 
-  private def site       = callsiteTyper.context
-  private def enclTrees  = site.enclosingContextChain map (_.tree)
-  private def enclPoses  = enclosingMacros map (_.macroApplication.pos) filterNot (_ eq NoPosition)
+  private lazy val site       = callsiteTyper.context
+  private lazy val enclTrees  = site.enclosingContextChain map (_.tree)
+  private lazy val enclPoses  = enclosingMacros map (_.macroApplication.pos) filterNot (_ eq NoPosition)
+
+  private def lenientEnclosure[T <: Tree : ClassTag]: Tree = enclTrees collectFirst { case x: T => x } getOrElse EmptyTree
+  private def strictEnclosure[T <: Tree : ClassTag]: T = enclTrees collectFirst { case x: T => x } getOrElse (throw new EnclosureException(classTag[T].runtimeClass, enclTrees))
```
Comment (paulp) [inline src/compiler/scala/reflect/macros/runtime/Enclosures.scala:16]: Pattern matching like this works now? Funny, I missed the memo. Neato.

Comment (paulp) [inline src/compiler/scala/reflect/macros/runtime/Enclosures.scala:16]: https://issues.scala-lang.org/browse/SI-6890

File: src/reflect/scala/reflect/macros/Enclosures.scala [hunk 2]
```diff
@@ -15,7 +15,7 @@ trait Enclosures {
   /** The tree that undergoes macro expansion.
    *  Can be useful to get an offset or a range position of the entire tree being processed.
    */
-  val macroApplication: Tree
+  def macroApplication: Tree
```
Comment (retronym) [inline src/reflect/scala/reflect/macros/Enclosures.scala:18]: Changing a `val` to `def` isn't source compatible (a client of the API might rely on the stability). I'm not sure how we should treat that for `scala-reflect`.

Comment (xeno-by) [inline src/reflect/scala/reflect/macros/Enclosures.scala:18]: Sure, but the PR is against `master`. Would be nice to have something like a deprecation warning, but I'm not sure how to express that.

Comment (retronym) [inline src/reflect/scala/reflect/macros/Enclosures.scala:18]: Oh sorry, I got my wires crossed.

**Comment by paulp** [comment]
Reviewing from phone but lgtm.

**Comment by paulp** [comment]
You can't express it, but it is squarely in the category of "deprecating non-final" style situations which we must make deprecatable.

**Pull request merged**